pkg/backup/backup.go

```diff
@@ -174,94 +175,186 @@ func (b *Backup) Run() {
+	}
+
 	return rev, nil
 }
 
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
-	start := time.Now()
+// WriteSnap writes the snapshot to the backend.
+func WriteSnap(version string, rev int64, be backend.Backend, snap *Snap) (int64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
+	rc, err := snap.Snapshot(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get snapshot (%v)", err)
+	}
+
+	n, err := be.Save(version, rev, rc)
+	rc.Close()
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+// GetSnap gets the snapshot from the cluster member with the maximum revision.
+func GetSnap(kclient kubernetes.Interface, tls *tls.Config, namespace, clusterName string) (*Snap, int64, error) {
+	pods, err := getRunningPods(kclient, namespace, clusterName)
+	if err != nil {
+		return nil, 0, fmt.Errorf("failed to retrieve running etcd pods: (%v)", err)
+	}
+	if len(pods) == 0 {
+		return nil, 0, fmt.Errorf("no running etcd pods found")
+	}
+
+	m, rev := getMemberWithMaxRev(pods, tls)
+	if m == nil {
+		return nil, 0, fmt.Errorf("no reachable member")
+	}
+
+	return NewSnap(m.ClientURL(), tls), rev, nil
+}
 
+func NewSnap(url string, tls *tls.Config) *Snap {
 	cfg := clientv3.Config{
-		Endpoints:   []string{m.ClientURL()},
+		Endpoints:   []string{url},
 		DialTimeout: constants.DefaultDialTimeout,
-		TLS:         b.etcdTLSConfig,
+		TLS:         tls,
 	}
-	etcdcli, err := clientv3.New(cfg)
+	return &Snap{cfg: cfg}
+}
+
+type Snap struct {
+	cfg clientv3.Config
+	// mu protects lazy creation of cli.
+	mu sync.Mutex
+	// cli is lazily created when either Snapshot or Version functions
+	// are being called.
+	cli *clientv3.Client
+}
+
+// Snapshot returns a snapshot from the etcd member.
+func (s *Snap) Snapshot(ctx context.Context) (io.ReadCloser, error) {
+	cli, err := s.lazyClient()
+	if err != nil {
+		return nil, err
+	}
+	rc, err := cli.Maintenance.Snapshot(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to create etcd client (%v)", err)
+		return nil, fmt.Errorf("failed to get member snapshot (%v)", err)
 	}
-	defer etcdcli.Close()
+	return rc, nil
+}
 
+// Version returns the etcd version of the etcd member.
+func (s *Snap) Version() (string, error) {
+	cli, err := s.lazyClient()
+	if err != nil {
+		return "", err
+	}
 	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
-	resp, err := etcdcli.Maintenance.Status(ctx, m.ClientURL())
+	resp, err := cli.Maintenance.Status(ctx, s.cfg.Endpoints[0])
 	cancel()
 	if err != nil {
-		return err
+		return "", fmt.Errorf("failed to get member status (%v)", err)
 	}
+	return resp.Version, nil
+}
 
-	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
-	rc, err := etcdcli.Maintenance.Snapshot(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to receive snapshot (%v)", err)
+func (s *Snap) lazyClient() (*clientv3.Client, error) {
```

Comment:
do not expose implementation detail through method naming. 

also why client is a method on snap?

Reply:
`lazyClient()` is use to create client lazily and is used in both `Version()` and `Snapshot()`. So I refactor the common client initialization code as `lazyClient()`.

Reply:
i do not think the snap struct makes a lot of sense to me at the first place.

things are getting too confusing here.

first backup is the controller that manages backups. to make it less confusing, the actual backup is called snapshot.

now we are introducing two similar things again: snap and snapshot.

a backup can get a snap to get snapshots... it is just getting super confusing.